modules/pam_mkhomedir/mkhomedir_helper.c

```diff
@@ -26,275 +26,346 @@
+   free(newsource);
 
-	 if (res == 0)
-	     continue;
+   return retval;
+}
 
-	 if (res > 0) {
-	     if (pam_modutil_write(destfd, remark, res) == res)
-		continue;
-	 }
+/* Do the actual work of creating a home dir */
+static int
+create_homedir(struct dir_spec *parent, const struct passwd *pwd,
+	       mode_t dir_mode, const char *source, const char *dest)
+{
+   DIR *d = NULL;
+   struct dirent *dent;
+   struct dir_spec base;
+   int retval = PAM_SESSION_ERR;
 
-	 /* If we get here, pam_modutil_read returned a -1 or
-	    pam_modutil_write returned something unexpected. */
-	 pam_syslog(NULL, LOG_DEBUG, "unable to perform IO: %m");
-	 close(srcfd);
-	 close(destfd);
-	 closedir(d);
+   /* Create the new directory */
+   if (mkdirat(parent->fd, dest, 0700))
+   {
+      if (errno == EEXIST)
+	 return PAM_SUCCESS;
+      pam_syslog(NULL, LOG_ERR, "unable to create directory %s/%s: %m",
+		 parent->path, dest);
+      return PAM_PERM_DENIED;
+   }
 
-	 free(newsource);
-	 free(newdest);
+   if (dir_spec_at(&base, parent, dest) < 0)
+   {
+      retval = PAM_PERM_DENIED;
+      goto go_out;
+   }
 
-	 return PAM_PERM_DENIED;
-      }
-      while (res != 0);
-      close(srcfd);
-      close(destfd);
+   /* See if we need to copy the skel dir over. */
+   if ((source == NULL) || (strlen(source) == 0))
+   {
+      retval = PAM_SUCCESS;
+      goto go_out;
+   }
 
-      free(newsource);
-      free(newdest);
+   /* Scan the directory */
+   d = opendir(source);
+   if (d == NULL)
+   {
+      pam_syslog(NULL, LOG_DEBUG, "unable to read directory %s: %m", source);
+      retval = PAM_PERM_DENIED;
+      goto go_out;
+   }
 
+   for (dent = readdir(d); dent != NULL; dent = readdir(d))
+   {
+      /* Skip some files.. */
+      if (strcmp(dent->d_name,".") == 0 ||
+	  strcmp(dent->d_name,"..") == 0)
+	 continue;
+
+      retval = copy_entry(&base, pwd, dir_mode, source, dent);
+      if (retval != PAM_SUCCESS)
+	 goto go_out;
    }
-   closedir(d);
 
    retval = PAM_SUCCESS;
 
  go_out:
+   if (d != NULL)
+      closedir(d);
 
-   if (chmod(dest, 0777 & (~u_mask)) != 0 ||
-       chown(dest, pwd->pw_uid, pwd->pw_gid) != 0)
+   if (fchmodat(parent->fd, dest, dir_mode, AT_SYMLINK_NOFOLLOW) != 0 ||
+       fchownat(parent->fd, dest, pwd->pw_uid, pwd->pw_gid,
+		AT_SYMLINK_NOFOLLOW) != 0)
    {
       pam_syslog(NULL, LOG_DEBUG,
-		 "unable to change perms on directory %s: %m", dest);
-      return PAM_PERM_DENIED;
+		 "unable to change perms on directory %s: %m", base.path);
+      retval = PAM_PERM_DENIED;
    }
 
+   dir_spec_close(&base);
+
    return retval;
 }
 
 static int
-create_homedir_helper(const struct passwd *_pwd,
+create_homedir_helper(const struct passwd *_pwd, mode_t home_mode,
 		      const char *_skeldir, const char *_homedir)
 {
    int retval = PAM_SESSION_ERR;
+   struct dir_spec base;
+   char *cp = strrchr(_homedir, '/');
 
-   retval = create_homedir(_pwd, _skeldir, _homedir);
+   if (cp == NULL)
```

Comment:
Apparently, `cp` cannot be NULL here because of `pwd->pw_dir[0] != '/'` check added in the calling function.
Maybe moving that check and the code that follows it into this function would result to a more readable code.

Reply:
That's true. We've kept the check here just to be cautious with strrchr return value dereference without NULL check. It can be removed and I've done so.